### PR TITLE
Adds emag functions to chem dispensers

### PR DESCRIPTION
**Pull request by PKPenguin321**
When emagged, it will unlock new chemicals.

Standard chemistry chem dispensers will get:
- Space drugs
- Morphine
- Carpotoxin
- ~~Regenerative materia~~
- Miner's Salve (a chem that heals minor brute and burn while making the subject believe they are healthy)
- Generic toxin

Bar booze dispensers get:
- Ethanol
- Iron
- Mint toxin
- Atomic Bomb

Bar cola dispensers get:
- Thirteen loko
- Whiskey Cola
- ~~Morphine~~
- Mindbreaker
- Tirizene

The mutagen dispensers that afaik only exist in certain lavaland ruins get plasma.

_Known issues:_
- ~~Attacking a chem dispenser with an emag properly displays the emag message, but still displays the "You can't insert that into the chem dispenser!" message~~
- ~~Regenerative materia's name is too long and breaks the chem dispenser's formatting a bit (also it's a blob chem and using it might be kind of hacky, might want to move the emag version to its own chem)~~
- ~~You can emag them multiple times despite the check, however this will only display the message multiple times and will not add more buttons past the first emag~~
- Constructable R&D chem dispensers are ~~currently untested~~ tested. They gain all the emag chems regardless of their upgraded parts level. I am fine with this for now, but if it proves to be some kind of balance issue, it can be very easily removed later.

Maybe todo?:
- Add/remove chems based on feedback to this PR
- Make emagged chems require more of the chem dispenser's power to dispense **Edit:** Will do in future if this somehow proves to be unbalanced (I doubt it)
- Add a check to make sure the dispenser has an emagged_chems list before emagging to futureproof any possible new chem dispensers that lack an emagged_chems list (Maybe already done? A check exists, but I don't know if it works.) **Edit:** Apparently not necessary?

:cl: PKPenguin321
rscadd: You can now emag chemical dispensers, such as the ones in chemistry or the bar, to unlock illegal chemicals.
/:cl:

Merge token: https://github.com/tgstation/tgstation/pull/17965

### CODE REVIEW

**Comment by Iamgoofball** [comment]
Replace the regen materia with miner's salve, it's also saying you have full health memes

**Comment by PKPenguin321** [comment]
good idea goof

**Comment by Zilenan91** [comment]
@PKPenguin321 Giving the bartender easy access to Morphine is a very very very bad idea. He'll use his damp rag to instantly ggnore people with neurotoxin. The only reason it's fine right now is because he has to ask for it from chemistry, this makes him totally self-sufficient.

Other than the bar morphine :+1:

**Comment by Iamgoofball** [comment]
>Whiskey Cola

Why? You can already make this super easy its just whiskey and space cola

**Comment by PKPenguin321** [comment]
It has a decently high alcohol level and is cola themed which matches the vendor

If you have a better idea hit me up

**Comment by Iamgoofball** [comment]
Slurm

**Comment by Iamgoofball** [comment]
Brawndo

**Comment by Iamgoofball** [comment]
Both

**Comment by xxalpha** [comment]
I am not up to date on chemistry, but carpotoxin? Isn't dispensing this OP?

**Comment by Bawhoppen** [comment]
Yea everything else is okay, but Carpotoxin is super OP, don't give them that for free

**Comment by WJohn** [comment]
Emags are so impossibly useful and necessary at this point they should practically be something all traitors can purchase for free (one time).

**Comment by GunHog** [comment]
Everything here is fantastic, but please do not give them Carpotoxin! That can be made into zombie powder, better known as "parapen juice"!

**Comment by PKPenguin321** [comment]
That's the idea

**Comment by Iamgoofball** [comment]
stop being pussies, medbay has better ways to treat poisonings than ever

**Comment by GunHog** [comment]
> stop being pussies, medbay has better ways to treat poisonings than ever

Who do you think is going to be the **source** of these poisonings?

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 1]
```diff
@@ -43,6 +43,13 @@
 		"bromine",
 		"stable_plasma"
 	)
+	var/list/emagged_reagents = list(
+		"space_drugs",
+		"morphine",
+		"carpotoxin",
+		"regenerative_materia",
```
Comment (ChangelingRain) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: please, please, please do not put blob chems in anything but the blob, they're not meant to ever be analyzed or messed with

Comment (Iamgoofball) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: yeah he's gonna swap it for miner's salve dont worry fam

**Comment by Zilenan91** [comment]
@PKPenguin321 Please remove the morphine from the bar it is GOING to be bad and horribly abused.

**Comment by kevinz000** [comment]
Op as fuck regardless

**Comment by PKPenguin321** [comment]
**In this commit:**

*Chem changes:*
- Regenerative materia removed from chem dispensers
- Regenerative materia replaced with miners salve, a chemical that heals slight burn and brute but makes the patient believe they are at full health
- Morphine removed from soda dispenser
- Mindbreaker added to soda dispenser
- Tirizene added to soda dispenser (does stamina damage very quickly, and i believe it caps at 55, so not a full stun)
- Atomic Bomb added to booze dispenser (potent alcohol that induces drugginess and if too much is taken, eventual sleep)

*Code changes:*
- Extra check to stop the "You can't load the emag into the machine!" message
- Actual emag code shuffled a bit to try to fix the multiple emag messages bug, to no avail.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 2]
```diff
@@ -65,6 +72,13 @@
 	if(energy != oldenergy)
 		use_power(2500)
 
+/obj/machinery/chem_dispenser/emag_act(mob/user)
+	if(emagged || !emagged_reagents)
+		user << "<span class='warning'>\The [src] has no functional safeties to emag.</span>"
+		return
+	user << "<span class='notice'>You short out \the [src]'s safeties.</span>"
+	dispensable_reagents |= emagged_reagents//add the emagged reagents to the dispensable ones
```
Comment (PKPenguin321) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: if somebody could help me with this:

my check on line 76 is supposed to return the code if the machine is already emagged or if it has a null emagged_reagents list. apparently it doesn't work, as emagging the machine multiple times will throw the message on line 79 multiple times (however it won't add infinite buttons to the chem dispenser so that's nice).

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 3]
```diff
@@ -165,7 +179,7 @@
 			icon_beaker = image('icons/obj/chemical.dmi', src, "disp_beaker") //randomize beaker overlay position.
 		icon_beaker.pixel_x = rand(-10,5)
 		overlays += icon_beaker
-	else if(user.a_intent != "harm")
+	else if(user.a_intent != "harm" && !istype(I, /obj/item/weapon/card/emag))
 		user << "<span class='warning'>You can't load \the [I] into the machine!</span>"
```
Comment (PKPenguin321) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: this check on line 182 might be kinda snowflake but it works. if you have a better solution please tell me

**Comment by Zilenan91** [comment]
:+1:

**Comment by AdamElTablawy** [comment]
Tirizine, last time I checked, does not cap. It's super strong, strong enough to put someone in stamina crit (unconcious) extremely quickly and for a long time. No tirizine. Perhaps chloral instead.

I disagree with adding yet more to the emag. I'd rather if this was the result of hacking the machines with a multitool and screwdriver rather than an emag.

Also, nono to carpotoxin. It's necessary to make zombie powder with it, which was the shit in old parapen slash parasting, and the only thing keeping that balanced is how rare it is. If you do that, you will very very quickly see a return of what amounted to the old parapen, except now in spraybottle/smoke grenade form.

**Comment by PKPenguin321** [comment]
Zombie powder itself isn't really overpowered. You might be new so let me educate you on its removal:
Parapen+C4 was the truly overpowered thing. The parapen came with iirc either 30 or 50 units of zombie powder (an extraordinarily large amount) and in an extremely convenient delivery package (the pen, which could be easily jabbed into anyone with an extremely small and hard to spot sprite). C4 could be purchased, set, applied, and detonated on the target in the time it took for the powder to wear off, and was a guaranteed gibbing. Not just an instakill like C4 today, but a gibbing, and with no chance of dropping a brain or any trace aside from an ID. It was the sum of these factors that was unfun-overpowered, not zombie powder itself. 

Zombie powder right now is effectively removed. It requires carpotoxin which requires carp meat or koi beans, and outputs *10x less* zombie powder than the ingredients put in (100 morphine + 100 carpotoxin + 100 copper = 10u zombie powder). Since parapens were nerfed, I've seen 4u of zombie powder ever brewed, and that was by myself for fun so I could fake my death on the shuttle for 5 seconds.

Zombie powder is effectively a stun that lasts as long as it is in your system, just like neurotoxin (which is extraordinarily easier to get). It also mutes, not unlike mute toxin or that other chemical that mutes and does losebreath (both of which are extremely common compared to zombie powder). It also has the side effect of making you appear dead, which opens the door for all kinds of death-faking shenanigans, which is why I'm so fond of it. (It also does extremely minor toxin damage.)

tldr: Zombie powder is not OP. It is pretty much weaker than neurotoxin (the bar drink) due to how rare it is, which is a shame because of how cool of a chem it is.

As for tirizine, I'll look into it not having a cap for you. I'll test it again, but I'm 99% sure it does have a cap. It is only really dangerous if you have 45 of any other damage, as that + the stamina would make you get knocked down.

**Comment by Iamgoofball** [comment]
to add onto zombie powder:

The stats page says in the last 30 days, **22** units have been made.

**Comment by xxalpha** [comment]
And that's without dispensers freely giving it to you.

**Comment by Sarcalogo** [comment]
Why not go all the way and add Venom, Amantin, etc. Or add some other reagents that when combined, can make those chems. Like add a "neon, helium, cobalt" that can be used to make some other traitor chems.

Also Miner's salve can be created already, not sure why you bothering to put it in dispenser. If anything, add an oil dispense button because it's used more often.

**Comment by GunHog** [comment]
>add an oil dispense button because it's used more often.

T4 Portable dispensers have this, however given the R&D nerf, it is going to be a rare sight.

**Comment by Iamgoofball** [comment]
@Sarcalogo no mixable traitor chems

**Comment by PKPenguin321** [comment]
requesting code review/help with the issues i pointed out

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 2 above)
Comment (GunHog) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Maaaaybe because you did not set the `emagged` var, and the `emagged_reagents` var is a list and cannot be checked properly in the manner you provided. If you set the `emagged` var, you do not need to check `emagged_reagents` anyway.

Comment (PKPenguin321) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: ohhh shit i'm an idiot, i figured emag_act set emagged automatically (despite me never even calling parent)

the check for emagged_reagents is to ensure the machine actually has an emagged_reagents list but I could remove it if it isn't needed

Comment (GunHog) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: You should not need to check that the `emagged_reagents` list has elements in it, because you have no code that modifies it. If you are planning to allow the list to change or be empty during gameplay, then checking that it has entries via `emagged_reagents.len`. If it has a length greater than 0, it has entries.

**Comment by Lumipharon** [comment]
I am opposed to this simply because fuck emags being useful in like everysituationeveralwaysgottabuyemag.

**Comment by PKPenguin321** [comment]
i actually havent bought an emag in ages personally

**Comment by PKPenguin321** [comment]
Just tested tirizene again, it does in fact cap out at 55 stamina damage. Enough to put you at orange health speed, not enough to knock you over.

Commit coming in a second that should make this good to go.

**Comment by optimumtact** [comment]
Bringing back parapen since 1801
